### PR TITLE
Fix `ScalarValue::to_array_of_size` for DenseUnion

### PR DESCRIPTION
**Pull request by kylebarron**
## Which issue does this PR close?

Closes https://github.com/apache/datafusion/issues/13762

## Rationale for this change

Fix support for dense union.

## What changes are included in this PR?

Fix `ScalarValue::to_array_of_size` for DenseUnion

## Are these changes tested?

Yes. The test from https://github.com/apache/datafusion/pull/13777 is uncommented.

### CODE REVIEW

**Comment by kylebarron** [comment]
I also validated in https://github.com/geoarrow/geoarrow-rs/pull/951 that this is now working for storing geospatial geometries in a dense union!

Mirroring the [PostGIS `ST_GeomFromText`](https://postgis.net/docs/ST_GeomFromText.html):

```
SELECT ST_GeomFromText('LINESTRING(-71.160281 42.258729,-71.160837 42.259113,-71.161144 42.25932)');
```

```
---- udf::native::io::wkt::test::test stdout ----
+----------------------------------------------------------------------------------------------------+
| st_geomfromtext(Utf8("LINESTRING(-71.160281 42.258729,-71.160837 42.259113,-71.161144 42.25932)")) |
+----------------------------------------------------------------------------------------------------+
| {=[{x: -71.160281, y: 42.258729}, {x: -71.160837, y: 42.259113}, {x: -71.161144, y: 42.25932}]}    |
+----------------------------------------------------------------------------------------------------+
```

Under the hood this is stored as a [rather complex dense union](https://github.com/geoarrow/geoarrow/blob/main/format.md#geometry), but allows for any combination of geometry type and dimension with a unique DataType that's streaming-friendly.

File: datafusion/common/src/scalar/mod.rs [hunk 1]
```diff
@@ -2453,15 +2453,23 @@ impl ScalarValue {
                             value.to_array_of_size(size)?
                         } else {
                             let dt = field.data_type();
-                            new_null_array(dt, size)
+                            match mode {
+                                UnionMode::Sparse => new_null_array(dt, size),
+                                // In a dense union, only the child with values needs to be
```
Comment (alamb) [inline datafusion/common/src/scalar/mod.rs:2458]: I double checked the spec and I agree with this check: https://arrow.apache.org/docs/format/Columnar.html#dense-union

File: datafusion/common/src/scalar/mod.rs [hunk 2]
```diff
@@ -5642,14 +5650,12 @@ mod tests {
             // null array
             Arc::new(NullArray::new(3)),
             // dense union
-            /* Dense union fails due to https://github.com/apache/datafusion/issues/13762
             {
```
Comment (alamb) [inline datafusion/common/src/scalar/mod.rs:5653]: 🎉

**Review by alamb — APPROVED** [review]
Thank you @kylebarron  -- this looks great to me

**Pull request merged**